pkg/kubectl/cmd/util/cmdtools.go

```diff
@@ -0,0 +1,265 @@
+/*
+Copyright 2015 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"math/rand"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+	"strings"
+
+	"github.com/golang/glog"
+
+	"k8s.io/kubernetes/pkg/kubectl/util/term"
+)
+
+const (
+	// TODO: on Windows rely on 'start' to launch the editor associated
+	// with the given file type. If we can't because of the need of
+	// blocking, use a script with 'ftype' and 'assoc' to detect it.
+	defaultEditor = "vi"
+	defaultShell  = "/bin/bash"
+	windowsEditor = "notepad"
+	windowsShell  = "cmd"
+	gitDiff       = "git diff --no-index"
+	normalDiff    = "diff"
+)
+
+type CmdName string
+
+const (
+	EditorCmd CmdName = "editor"
+	DiffCmd   CmdName = "diff"
+)
+
+// CmdTools include
+// editor and diff tools
+type CmdTool struct {
+	Name  CmdName
+	Args  []string
+	Shell bool
+}
+
+// NewDefaultCmdTool creates a struct CmdTool that uses the OS environment to
+// locate the editor or diff program, looking at environment variable to find
+// the proper command line. for editor: If the provided editor has no spaces,
+// or no quotes, it is treated as a bare command to be loaded.
+// Otherwise, the string will be passed to the user's shell for execution.
+func NewDefaultCmdTool(name CmdName, envs []string) CmdTool {
+	exec, shell := defaultEnv(name, envs)
+	return CmdTool{
+		Name:  name,
+		Args:  exec,
+		Shell: shell,
+	}
+}
+
+func defaultEnvShell() []string {
+	shell := os.Getenv("SHELL")
+	if len(shell) == 0 {
+		shell = platformize(defaultShell, windowsShell)
+	}
+	flag := "-c"
+	if shell == windowsShell {
+		flag = "/C"
+	}
+	return []string{shell, flag}
+}
+
+func defaultEnv(cmdName CmdName, envs []string) ([]string, bool) {
+	var cmd string
+	for _, env := range envs {
+		if len(env) > 0 {
+			cmd = os.Getenv(env)
+		}
+		if len(cmd) > 0 {
+			break
+		}
+	}
+	if len(cmd) == 0 {
+		switch cmdName {
+		case EditorCmd:
+			cmd = platformize(defaultEditor, windowsEditor)
+		case DiffCmd:
+			cmd = checkGitDiffInstalled(gitDiff, normalDiff)
+		}
+	}
+
+	if !strings.Contains(cmd, " ") {
+		return []string{cmd}, false
+	}
+	if !strings.ContainsAny(cmd, `"'\`) {
+		return strings.Split(cmd, " "), false
+	}
```

Comment:
Something like this:

```bash
export KUBECTL_EXTERNAL_DIFF="git
diff
--no-index"

echo '$KUBECTL_EXTERNAL_DIFF = '$KUBECTL_EXTERNAL_DIFF

kubectl create deployment hello-world --image nginx
kubectl get deployment hello-world -o json > ./hello-world-deployment.json
kubectl apply diff-last-applied pods -f ./hello-world-deployment.json
```

Output:

```
$KUBECTL_EXTERNAL_DIFF = git diff --no-index
Error from server (AlreadyExists): deployments.extensions "hello-world" already exists
error: unable to launch the diff "git\ndiff\n--no-index"
```

Running `"bash -c $KUBECTL_EXTERNAL_DIFF"` in this script works just fine; it's only broken due to the custom string split code here in this function. Now obviously this is a bit of a contrived example. My point is that the shell is an extremely complicated system and has a million edge cases.

There's two ways to fix this:

1. Add '\n' to the list of things we check the string for and update the string.Split() code to split on '\n' as well.
2. Punt to the shell in all cases instead of trying to parse the string ourselves.

I would strongly prefer punting to the shell in all cases-- offload more work to the shell.
Does that make sense?